src/server/scripts/EasternKingdoms/SunkenTemple/boss_avatar_of_hakkar.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ScriptMgr.h"
+#include "sunken_temple.h"
+#include "InstanceScript.h"
+#include "ScriptedCreature.h"
+#include "SpellInfo.h"
+#include "Player.h"
+
+enum Events
+{
+    EVENT_CAST_LASH              = 1,
+    EVENT_CAST_CAUSE_INSANITY,
+    EVENT_CAST_CURSE_OF_TONGUES,
+    EVENT_SUPPRESSOR
+};
+
+enum Spells
+{
+    SPELL_LASH             = 6607,
+    SPELL_CAUSE_INSANITY   = 12888,
+    SPELL_CURSE_OF_TONGUES = 12889,
+};
+
+enum Yells
+{
+    SAY_SPAWN_AVATAR = 0,
+    SAY_AGRO_AVATAR  = 1
+};
+
+enum DisplayID
+{
+    DISPLAY_SHADE_OF_HAKKAR  = 7690,
+    DISPLAY_AVATAR_OF_HAKKAR = 8053
+};
+
+
+/*#####
+# boss_avatar_of_hakkar
+#####*/
+
+class boss_avatar_of_hakkar : public CreatureScript
+{
+public:
+    boss_avatar_of_hakkar() : CreatureScript("boss_avatar_of_hakkar") { }
+
+    struct boss_avatar_of_hakkarAI : public BossAI
+    {
+        boss_avatar_of_hakkarAI(Creature* creature) : BossAI(creature, BOSS_AVATAR_OF_HAKKAR) { }
+
+        void JustEngagedWith(Unit* who) override
+        {
+            BossAI::JustEngagedWith(who);
+            Talk(SAY_AGRO_AVATAR);
+            events.ScheduleEvent(EVENT_CAST_LASH, 4s, 7s);
+            events.ScheduleEvent(EVENT_CAST_CAUSE_INSANITY, 6s, 14s);
+            events.ScheduleEvent(EVENT_CAST_CURSE_OF_TONGUES, 14s, 21s);
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (!UpdateVictim())
+                return;
+
+            events.Update(diff);
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            while (uint32 eventId = events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                case EVENT_CAST_LASH:
+                    DoCastVictim(SPELL_LASH);
+                    events.ScheduleEvent(EVENT_CAST_LASH, 4s, 7s);
+                    break;
+                case EVENT_CAST_CAUSE_INSANITY:
+                    if (CheckNearbyPlayers())
```

Comment:
What is the reason for this check?

Reply:
I needed this when I was using `DoCastVictim` until I learned about` SelectTarget(SelectTargetMethod::Random, 0, NonTankTargetSelector(me))`. Since, if the **Cause Insanity** skill was used on a single player, the boss's aggro was reset.